### PR TITLE
Add cloudwatch LogAppender

### PR DESCRIPTION
**Pull request by kusumachalasani**
Create a cloudwatch logAppender to send logs to cloudwatch if access and secret keys are shared through the config.  AWS details are passed as part of dbconfig.
"logging": {
    "cloudwatch": {
      "accessKeyId": "",
      "logGroup": "",
      "region": "",
      "secretAccessKey": ""
    },
    "type": "null"
  }

### CODE REVIEW

File: src/main/java/com/autotune/operator/KruizeDeploymentInfo.java [hunk 1]
```diff
@@ -65,6 +65,14 @@ public class KruizeDeploymentInfo {
     public static String database_admin_username;
     public static String database_admin_password;
     public static String database_ssl_mode;
+
+    public static String cloudwatch_logs_access_key_id;
+    public static String cloudwatch_logs_secret_access_key;
+    public static String cloudwatch_logs_log_group;
+    public static String cloudwatch_logs_region;
+    public static String cloudwatch_logs_log_level;
+    public static String cloudwatch_logs_log_stream;
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/operator/KruizeDeploymentInfo.java:74]: Can you also please make this variable set via env 
you can do that by 
 declaring  it in KruizeConstants.KRUIZE_CONFIG_ENV_NAME
 
 public static final String CLOUDWATCH_LOGS_ACCESS_KEY_ID = "cloudwatch_accesskeyid";  
 make sure to have only one Underscore in value 


and in crc yaml in ConfigMap -> in kruizeconfigjson

declare like this
"cloudwatch" : {
         "accesskeyid" : "",
         "secretaccesskey:"",   etc..
}

File: src/main/java/com/autotune/utils/CloudWatchLogSender.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    }
+    public static void configureCloudWatchLog() {
+
+        // Check if access details are provided
+        if (cloudwatch_logs_access_key_id != null && !cloudwatch_logs_access_key_id.isEmpty() && cloudwatch_logs_secret_access_key != null && !cloudwatch_logs_secret_access_key.isEmpty() && cloudwatch_logs_region != null && !cloudwatch_logs_region.isEmpty()) {
+            try {
+                // Define default values for attributes if they are empty or null
+                String cw_logs_log_group = cloudwatch_logs_log_group == null || cloudwatch_logs_log_group.isEmpty() ? "kruize-logs" : cloudwatch_logs_log_group;
+                String cw_logs_log_stream = cloudwatch_logs_log_stream == null || cloudwatch_logs_log_stream.isEmpty() ? "kruize-stream" : cloudwatch_logs_log_stream;
+                String cw_logs_log_level = cloudwatch_logs_log_level == null || cloudwatch_logs_log_level.isEmpty() ? "INFO" : cloudwatch_logs_log_level;
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/CloudWatchLogSender.java:34]: Can cw_logs_log_level and existing LOGGING_LEVEL  env can have different values ?

File: src/main/java/com/autotune/utils/CloudWatchLogSender.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+                // Create a CloudWatchLogsClient with provided credentials and region
+                CloudWatchLogsClient logsClient = CloudWatchLogsClient.builder()
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/CloudWatchLogSender.java:37]: should be surrounded by try and catch. even if cloud watch fails application should run. But needs to be reported

File: src/main/java/com/autotune/utils/CloudWatchLogSender.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+                // Create a configuration builder
+                ConfigurationBuilder<BuiltConfiguration> builder = ConfigurationBuilderFactory.newConfigurationBuilder();
+                LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
+
+                // Define the console appender
+                AppenderComponentBuilder consoleAppenderBuilder = builder.newAppender("consoleLogger", "Console")
+                        .addAttribute("target", "SYSTEM_OUT")
+                        .add(builder.newLayout("PatternLayout")
+                                .addAttribute("pattern", "%d{yyyy-MM-ddHH:mm:ss.SSS} %level [%t][%F(%L)]-%msg%n"));
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/CloudWatchLogSender.java:60]: can this moved to constant and should be same as existing log handler

File: src/main/java/com/autotune/utils/KruizeConstants.java [hunk 5]
```diff
@@ -405,6 +412,21 @@ public static final class DATABASE_ENV_NAME {
         public static final String DATABASE_SSL_MODE = "database_sslmode";
     }
 
+    /**
+     * In order to assign values to the static variables of KruizeDeploymentInfo
+     * using Java reflection, the class variables are utilized, and therefore,
+     * if any new variables are added, their corresponding declaration is necessary.
+     * Ref InitializeDeployment.setConfigValues(KruizeConstants.CONFIG_FILE, KruizeConstants.KRUIZE_CONFIG_ENV_NAME.class);
+     */
+    public static final class CLOUDWATCH_LOGS_ENV_NAME {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/KruizeConstants.java:421]: should be moved inside KRUIZE_CONFIG_ENV_NAME 
and value should have only one underscore

File: src/main/java/com/autotune/utils/CloudWatchLogSender.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    private static void createLogStream(CloudWatchLogsClient logsClient, String logGroupName, String logStreamName) {
+        CreateLogStreamRequest request = CreateLogStreamRequest.builder()
+                .logGroupName(logGroupName)
+                .logStreamName(logStreamName)
+                .build();
+        logsClient.createLogStream(request);
+        LOGGER.info("Created log stream: {} in log group: {}", logStreamName, logGroupName);
+    }
+}
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/CloudWatchLogSender.java:128]: EOF missing

File: src/main/java/com/autotune/utils/CustomAppender.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            this.level = level;
+        }
+
+        @Override
+        public Result filter(LogEvent event) {
+            return event.getLevel().isMoreSpecificThan(level) ? onMatch : onMismatch;
+        }
+    }
+
+}
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/utils/CustomAppender.java:113]: EOF missing

File: pom.xml [hunk 8]
```diff
@@ -88,6 +88,12 @@
             <scope>compile</scope>
         </dependency>
 
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>cloudwatchlogs</artifactId>
+            <version>2.17.102</version>
```
Comment (msvinaykumar) [inline pom.xml:94]: Please declare this into constant 2.17.102

**Comment by msvinaykumar** [comment]
@kusumachalasani "I'll submit another pull request to enable logging of all requests and responses at the debug level. For a temporary period until the resolution of https://issues.redhat.com/browse/KRUIZE-95, we'll configure Cloudwatcher to log at the debug level, and pod logs at the info level. Afterward, both Cloudwatcher and pod logs should revert to the info level.

File: src/main/java/com/autotune/operator/KruizeDeploymentInfo.java (hunk 1 above)
Comment (chandrams) [reply src/main/java/com/autotune/operator/KruizeDeploymentInfo.java:74]: @kusumachalasani - Can you please document the steps to setup and test this

File: src/main/java/com/autotune/utils/CloudWatchLogSender.java (hunk 2 above)
Comment (kusumachalasani) [reply src/main/java/com/autotune/utils/CloudWatchLogSender.java]: yes. CloudWatch logging level can be different from the console.

File: src/main/java/com/autotune/operator/KruizeDeploymentInfo.java (hunk 1 above)
Comment (kusumachalasani) [reply src/main/java/com/autotune/operator/KruizeDeploymentInfo.java:74]: Updated the kruizeconfig related design documents for the same.
To enable cloudwatch logs, we need to provide aws access / secret keys as mentioned in the document.

**Review by msvinaykumar — APPROVED** [review]
LGTM

File: src/main/java/com/autotune/operator/KruizeDeploymentInfo.java (hunk 1 above)
Comment (chandrams) [reply src/main/java/com/autotune/operator/KruizeDeploymentInfo.java:74]: @kusumachalasani - I don't see the doc changes in this PR

File: design/KruizeConfigurationProcedure.md [hunk 9]
```diff
@@ -48,6 +48,14 @@ data:
         "c3p0maxstatements": 50,
         "hbm2ddlauto": "update",
         "showsql": "true"
+      },
+      "cloudwatch": {
+        "accessKeyId": "",
```
Comment (chandrams) [inline design/KruizeConfigurationProcedure.md:53]: Update the environment variables section if these can be configured through env variables.

**Review by dinogun — APPROVED** [review]
LGTM

File: design/KruizeConfigurationProcedure.md (hunk 9 above)
Comment (chandrams) [reply design/KruizeConfigurationProcedure.md:53]: @kusumachalasani - Are the env variables same as that in the manifest to pass the cloudwatch keys?

**Pull request merged**